File: Top-Interview-Questions-Leetcode/Easy-Collections/math/count-primes.py
class Solution:
    def __init__(self):
        self.mx = None
        self.primes = []
    
    def __sieve(self):
        self.primes = [True for i in range(self.mx)]
        
        self.primes[0] = self.primes[1] = False
        
        i = 2
        
        while i * i < self.mx:
            if self.primes[i] == False:
                i += 1
                continue
            
            for j in range(i * i, self.mx, i):
                self.primes[j] = False
            
            i += 1
        
    def countPrimes(self, n: int) -> int:
        self.mx = n
        
        if n < 2:
            return 0
        
        self.__sieve()
        
        cnt = 0
        
        for i in range(2, n):
            if self.primes[i]:
                cnt += 1
        
        return cnt
        

File: Top-Interview-Questions-Leetcode/Easy-Collections/math/count-primes.cpp
#include <iostream>
#include <vector>

class Solution {
    
private:
    
    std::vector<bool> primes;
    int mx;

private:
    
    void sieve() {
        primes[0] = primes[1] = false;
        
        for (int i = 2; i * i < mx; ++i) {
            if (!primes[i]) continue;
            for (int j = i * i; j < mx; j += i) primes[j] = false;
        }
    }
    
public:
    int countPrimes(int n) {
        if (n == 0) return 0;
        
        primes.resize(n, true);
        mx = n;
        
        sieve();
        
        int cnt = 0;
        
        for (int i = 2; i < n; ++i) {
            if (primes[i]) cnt++;
        }
        
        return cnt;
    }
    
    
};